app/api/services/users/create_user.py

```diff
@@ -1,27 +1,20 @@
 from uuid import uuid4
 
-from api.db.models.user_models import Role, User
+from api.db.models.user_models import User
 from api.route.api_context import ApiContext
-from api.route.models.user import UserParams, UserResponse
 
 
 # TODO: separate controller and service concerns
 # https://github.com/navapbc/template-application-flask/issues/49#issue-1505008251
-def create_user(api_context: ApiContext) -> UserResponse:
-    request = UserParams.parse_obj(api_context.request_body)
+def create_user(user_data: dict, api_context: ApiContext) -> User:
+    # TODO: move this code to service and/or persistence layer
+    user = User(**user_data)
```

Comment:
Before I dig into this too much, are you planning to adjust this? I'd really like to not have dictionaries. In this case it's fine because the request model matches the DB model, but that won't always be the case for every API, how do we process that into a meaningful object?

Reply:
if the names don't match we'd use [serialization/deserialization keys](https://marshmallow.readthedocs.io/en/stable/quickstart.html#specifying-serialization-deserialization-keys) in the schema

Reply:
or is there a way to keep it as a dict for patch but use [post_load](https://marshmallow.readthedocs.io/en/stable/quickstart.html#deserializing-to-objects) for the other endpoints to convert to an object?

Reply:
I think any coupling of request to the DB schema should be seen as an ideal scenario, but not a required one. In this case where we're making up our own endpoint it's simple to make the request and DB models match, but that doesn't always hold up (eg. NJ where we didn't get to decide on how either was structured and 1 request maps to 20 tables in a different shape with 100 calculated params in-between).

Let's assume this fairly simple scenario. You have this POST request:
```
{
   "name": "bob",
   "age": 20,
   "favorite_colors": ["blue", "red"],
}
```
Now lets assume that the DB has a single table with the following columns (I know you wouldn't structure it this way):
- name: str
- age: int
- likes_blue: bool
- likes_red: bool

Ideally I'd be able to work with creating that model by doing something like:
```py
db_record = NameObject(
    name=request.name,
    age=request.age,
    likes_blue="blue" in request.favorite_colors,
    likes_red="red" in request.favorite_colors,
)
```
If request were a dictionary there would be a bit of issue with typing I'd imagine (I'm actually surprised it's okay with the patch endpoint approach we have as-is). And beyond the typing issue, it would be more error-prone/effort for development to have to look back at the model to figure out how the dict it validated is structured.

---

Serializing it into an object is good, but besides marshmallow dataclasses, is there any way to do that without re-declaring the model again (which might be fine, plus pydantic is convenient for loading from JSON in that case).

Reply:
with that example, the request schema would be something like

```
class PersonSchema(marshmallow.Schema):
  name = fields.String()
  age = fields.Integer()
  favorite_colors = fields.List(fields.String) # could be fields.Enum instead of fields.String but that's not relevant to our discussion
```

and then in # db.models.person we'd have 

```
class Person(Base):
  name = Column(Text)
  likes_blue = Column(Boolean)
  likes_red = Column(Boolean)
```

then in our controller we'd do
```
def create_person(person_request_data: dict):
  person = Person(
    name=person_request_data["name"],
    age=person_request_data["age"],
    likes_blue="blue" in person_request_data["favorite_colors"],
    likes_red="red" in person_request_data["favorite_colors"],
  )
```

and we'd still get runtime API request validation already from the Marshmallow schema, and we'd still get runtime db model constructor level validation with the SQLAlchemy model constructors.

@zelgadis yeah I started out having objects, using a marshmallow schema post_load hook like

```
class PersonSchema(Schema):
  ...
  @post_load
  def make_person(self, data: dict):
    return Person(
      name=data["name"],
      age=data["age"],
      likes_blue="blue" in data["favorite_colors"],
      likes_red="red" in data["favorite_colors"],
    )
```
but the issue is with the patch endpoint, where we want to allow clients to pass requests that don't have fields that are marked as required in the schema, and where we want to know exactly what fields were passed and not default them to None, since there's a semantic difference between PATCH set field X to null vs PATCH request that doesn't have field X. but if we convert it to an object in that case, the object will be an instance of the class, which means it will have all the attributes.

as such i think dictionaries with runtime type validation seems like the best approach.

Reply:
As a thought exercise, if i were to ask you to define a class `UserPatchParams` that represents the request data for `PATCH /v1/users/:user_id`, how would you define it?

Reply:
As a point of comparison, [this](https://github.com/navapbc/template-application-flask/blob/cdc883299e518d1c295c3b472c1e31a739bf690e/app/api/route/request.py) is what we needed to do in order to support PATCH requests with pydantic models. What it's doing is that a pydantic class will keep track of the fields that were set in a hidden field `__fields_set__`, and we use that to reverse engineer the dict that was originally used to create the instance of the pydantic model, but at that point I don't see how that's different from using a `dict` directly and just cutting out the middle step of turning it into an object first.

Taking a step back, here's the lifetime of the data
HTTP request (byte string) --> Flask framework (parse json to dict) --> App controller / route handler (runtime validation of request data) --> App service (work with domain objects)

And what we're discussing is whether we should add an extra layer of "convert request data to request data object" at the controller level.

Reply:
@lorenyu Okay gotcha, thanks for explaining! I see what value the dict is adding here and I would be okay with this approach. I still think I prefer the old approach of passing request data objects (eg `UserPatchParams`) to the service methods, and enumerating the `get_set_fields` in the PATCH case!

Reply:
> As a thought exercise, if i were to ask you to define a class `UserPatchParams` that represents the request data for `PATCH /v1/users/:user_id`, how would you define it?

This got my thinking cap on! I don't have a good sense of whether or not what I'm about to propose is a good idea or not, but I'm going to share anyways 😛 

In the `UserPatchParams` paradigm, we have fields marked with `Optional`, since they don't need to be specified through the API:

```
first_name: Optional[str]
```

This is essentially the same as:

```
first_name: Union[str, None]
```

I wonder if we could extend that to include an `Omitted` option as well? 

```
first_name: Union[str, None, Omitted]
```

In the constructor for the `UserPatchParams`, all fields could be `Omitted` by default, and only set to a value or `None` if they were included in the request. Then you could loop over the fields and set them only if they were not `Omitted`:

```
for attribute, value in params.__dict__.items():
  if type(value) == Omitted:
    # this attribute was not set in the request - skip it!
    continue

  setattr(user, attribute, value)
```

Reply:
I did have something in an earlier version of this to handle the "was it set in the request" problem. Basically, it re-references the raw-json to see if the value was present or not. Don't think it was robust enough for the heavily-nested objects, but worked in the simple case.

An idea that might overcomplicate things, but not sure if it'll work at all. Would it be possible to define the models in Pydantic, so we can use them like the existing approach, but export the pydantic schema (see: https://docs.pydantic.dev/usage/schema/) and somehow import that into marshmallow schema (see: https://marshmallow.readthedocs.io/en/stable/quickstart.html#creating-schemas-from-dictionaries ) - Would probably require some conversion utils, but maybe? Although for that, it'd just be easier to switch entirely to fastapi which is literally pydantic->openapi - https://fastapi.tiangolo.com/

Reply:
What type is the output of the `load` function? If it's still a dict, that doesn't help. My concern is that doing development with a dict is incredibly painful, I have to constantly refer back to the object manually and none of my IDE functionality works (auto-complete, type hints, and so on). In addition, `val = fields["some_param"]` is always going to run into typing issues.

If I started with a system that did it this way, the first thing I'd do is go create a pydantic model to load that JSON into as keeping that in sync with the request model is going to save me several magnitudes of dev-time compared to leaving it a raw dict.

Reply:
`load` returns a `dict`, but the amount of time it will "live" as a `dict` is basically nothing, since the service will immediately construct a domain object immediately from the request params i.e. either
```
def update_user(data):
  # validate data
  user = User(first_name=data["first_name"]...)
```
or
```
  user = User(**data)
```
so that's basically equivalent to converting it to a Pydantic model. but it removes the need to have another entire layer of classes in between.

So in my previous suggestion, the life cycle is "network json string" --> "dict" --> validate dict --> convert to domain object/ and in your suggestion it's adding an additional step in between so it's like "network json string" --> "dict" --> validate dict --> convert to request object --> convert to domain object. But my argument is that "convert to request object" isn't a valuable step since the service is immediately going to "convert to domain object" as the first first step of its operation.

Reply:
In this particular endpoint I agree, any additional conversion is basically redundant because the request model is the same as the DB model. But this is the happy-path case, this is how we'd love to setup most endpoints, but what about the complex scenario?

For example, the update application endpoint in Mass PFML. The `ApplicationRequestBody` needed a lot of parsing and pulling out of values in order to update various tables. Doing that with a dictionary would be quite painful / error prone.

It might be that the best approach if they were using this API structure would be to make a Pydantic model for this specific endpoint, but for other simpler endpoints skip that as it's not worthwhile. Does that seem reasonable? If so, could we add an example endpoint that does that to help show how you might do that?

Reply:
It seems like we're having difficulty reaching consensus on this. I tend to agree with @chouinar but I don't want to block forward progress.. Would it be helpful to spend 30 - 45 min discussing this live?

Some questions I have:
* What are we trying to gain by moving away from the `get_set_parameters` approach?
* Why does making Params classes for nested objects get complicated? Is it just that we have to create a Params class for each nested object?
* What is the "connexion magic" that we're trying to avoid?